src/dialer/index.js

```diff
@@ -105,22 +111,28 @@ class Dialer {
    *
    * @private
    * @param {PeerId|Multiaddr|string} peer - A PeerId or Multiaddr
-   * @returns {DialTarget}
+   * @returns {Promise<DialTarget>}
    */
-  _createDialTarget (peer) {
+  async _createDialTarget (peer) {
     const { id, multiaddrs } = getPeer(peer)
 
     if (multiaddrs) {
       this.peerStore.addressBook.add(id, multiaddrs)
     }
 
-    let addrs = this.peerStore.addressBook.getMultiaddrsForPeer(id) || []
+    let knownAddrs = this.peerStore.addressBook.getMultiaddrsForPeer(id) || []
 
     // If received a multiaddr to dial, it should be the first to use
     // But, if we know other multiaddrs for the peer, we should try them too.
     if (multiaddr.isMultiaddr(peer)) {
-      addrs = addrs.filter((addr) => !peer.equals(addr))
-      addrs.unshift(peer)
+      knownAddrs = knownAddrs.filter((addr) => !peer.equals(addr))
+      knownAddrs.unshift(peer)
+    }
+
+    const addrs = []
+    for (const a of knownAddrs) {
+      const resolvedAddrs = await this._resolve(a)
```

Comment:
resolve can throw, we shouldn't let a single address failure stop the whole dial, we need to catch it